src/librbd/mirror/snapshot/PromoteRequest.cc

```diff
@@ -246,18 +248,25 @@ void PromoteRequest<I>::handle_acquire_exclusive_lock(int r) {
 
 template <typename I>
 void PromoteRequest<I>::rollback() {
-  if (m_rollback_snap_id == CEPH_NOSNAP) {
+  uint64_t rollback_snap_id = CEPH_NOSNAP;
+  CephContext *cct = m_image_ctx->cct;
+
+  if (util::can_create_primary_snapshot(m_image_ctx, false, true, nullptr,
+                                        &rollback_snap_id)) {
```

Comment:
We are clearly missing a crucial integration test if such a serious bug wasn't caught.  Please add a test to `qa/workunits/rbd/rbd_mirror_snapshot.sh` that would fail on comparing image data when run against the current revision of the PR.